### PR TITLE
Support -H, -B, --jaeger for headers and baggage

### PR DESCRIPTION
**Pull request by kriskowal**

### CODE REVIEW

File: transport/http.go [hunk 1]
```diff
@@ -30,6 +30,8 @@ import (
 	"strconv"
 	"time"
 
+	"github.com/opentracing/opentracing-go"
+
```
Comment (prashantv) [inline transport/http.go]: nit: can remove blank since this is a third party import like the below import

File: bench_method_test.go [hunk 2]
```diff
@@ -144,10 +138,11 @@ func TestBenchmarkMethodCall(t *testing.T) {
 	}
 
 	tOpts := TransportOptions{
+		CallerName:  "bar",
 		ServiceName: "foo",
 		HostPorts:   []string{s.hostPort()},
 	}
-	tp, err := getTransport(tOpts, encoding.Thrift)
+	tp, err := getTransport(tOpts, encoding.Thrift, nil)
```
Comment (prashantv) [inline bench_method_test.go]: What is the difference in behaviour between `nil` here and `NoopTracer` in `bench_method.go`?

File: bench_method.go [hunk 3]
```diff
@@ -38,7 +39,7 @@ type benchmarkMethod struct {
 // up by making some number of requests through it.
 func (m benchmarkMethod) WarmTransport(opts TransportOptions, warmupRequests int) (transport.Transport, error) {
 	opts.benchmarking = true
-	transport, err := getTransport(opts, m.serializer.Encoding())
+	transport, err := getTransport(opts, m.serializer.Encoding(), opentracing.NoopTracer{})
```
Comment (prashantv) [inline bench_method.go]: Just realized, baggage only gets with the Jaeger tracer right? Perhaps we can't do this unless we explicitly disallow baggage with benchmarking? Otherwise, the initial request will use baggage but the benchmarking requests will not.

File: main.go [hunk 4]
```diff
@@ -252,8 +256,30 @@ func runWithOptions(opts Options, out output) {
...
 
+	if opts.TOpts.CallerName != "" {
+		if opts.TOpts.benchmarking {
+			out.Fatalf("Cannot override caller name when running benchmarks\n")
+		}
+	} else {
+		opts.TOpts.CallerName = "yab-" + os.Getenv("USER")
+	}
+
+	var tracer opentracing.Tracer
```
Comment (prashantv) [inline main.go]: can we move the logic of get a tracer into a separate method?

File: transport/interface.go [hunk 5]
```diff
@@ -23,15 +23,21 @@ package transport
...
-	Timeout time.Duration
-	Headers map[string]string
-	Body    []byte
+	Method          string
+	Timeout         time.Duration
+	Headers         map[string]string
+	Baggage         map[string]string
+	RoutingKey      string
+	RoutingDelegate string
+	ShardKey        string
```
Comment (prashantv) [inline transport/interface.go]: I don't know if routing key/delegate/shard key belong as top level fields on the request, unless we've made those transport-agnostic concerns in YARPC?

File: transport/interface.go [hunk 6]
```diff
@@ -23,15 +23,21 @@ package transport
...
+	Body            []byte
+	Tracer          opentracing.Tracer
```
Comment (prashantv) [inline transport/interface.go]: The `Request` and `Response` objects only contain values that represent the on-wire presentation currently. I'm not sure if Tracer should be here, should we move it into the `Call` interface instead?

File: CHANGELOG.md [hunk 7]
```diff
@@ -2,6 +2,9 @@ Changelog
 =========
 
 * Upgrade to TChannel 1.2.0 and add Open Tracing support.
+* Add support for -H key:value header arguments.
```
Comment (breerly) [inline CHANGELOG.md]: Suggest:

* Add support for headers with argument `-H key:value`.
* Add support for baggage with argument `-B key:value` when `--jaeger` is passed.

File: bench_method_test.go (hunk 2 above)
Comment (kriskowal) [inline bench_method_test.go]: If you thread nil, TChannel and YARPC will fall back to the global tracer. I’m going to explicitly thread NoopTracer and mandate a non-nil tracer in getTransport, so there’s no question for yab.

File: main_test.go [hunk 8]
```diff
@@ -384,8 +384,7 @@ func TestAlises(t *testing.T) {
 		{
 			args: []cmdArgs{
 				{"-P", "file"},
-				{"-H", "file"},
```
Comment (breerly) [inline main_test.go]: So we're stealing the `-H` flag for headers?

Comment (kriskowal) [inline main_test.go]: yes.

**Comment by kriskowal** [comment]
Added passing tests for baggage. Added a passing test for TChannel headers. Left a broken test for HTTP headers with #125 

Now seeking only a stamp from @prashantv

File: main.go [hunk 9]
```diff
@@ -325,5 +336,42 @@ func makeRequest(t transport.Transport, request *transport.Request) (*transport.
...
+		for k, v := range request.Baggage {
+			span = span.SetBaggageItem(k, v)
+		}
+		ctx = opentracing.ContextWithSpan(ctx, span)
+	}
+
 	return t.Call(ctx, request)
 }
+
+func makeInitialRequest(out output, transport transport.Transport, serializer encoding.Serializer, req *transport.Request) {
```
Comment (kriskowal) [inline main.go]: I am not sure where this new code is coming from. There might be rebase magic here.

File: transport.go [hunk 10]
```diff
@@ -29,23 +29,23 @@ import (
 	"io/ioutil"
 	"net"
 	"net/url"
-	"os"
 	"strings"
 
+	"github.com/opentracing/opentracing-go"
+	"github.com/uber/tchannel-go"
 	"gopkg.in/yaml.v2"
```
Comment (kriskowal) [inline transport.go]: third party group should be below first party group

File: transport/interface.go [hunk 11]
```diff
@@ -23,15 +23,19 @@ package transport
 import (
 	"time"
 
+	"github.com/opentracing/opentracing-go"
```
Comment (kriskowal) [inline transport/interface.go]: should be in same group as bellow

File: transport_test.go [hunk 12]
```diff
@@ -21,10 +21,13 @@
 package main
 
 import (
+	"fmt"
 	"os"
 	"testing"
 	"time"
 
+	"github.com/opentracing/opentracing-go"
+	opentracing_ext "github.com/opentracing/opentracing-go/ext"
```
Comment (kriskowal) [inline transport_test.go]: should separate third party group below

File: bench_method.go [hunk 13]
```diff
@@ -25,6 +25,7 @@ import (
 	"sync"
 	"time"
 
+	"github.com/opentracing/opentracing-go"
```
Comment (prashantv) [inline bench_method.go]: +1, same throughout this change

File: server_util_test.go [hunk 14]
```diff
@@ -78,11 +83,21 @@ func (methodsT) echo() handler {
 
 func (methodsT) traceEnabled() handler {
```
Comment (prashantv) [inline server_util_test.go]: can you add a comment that explains the returned result from the method

File: bench_method_test.go [hunk 15]
```diff
@@ -54,24 +55,18 @@ func benchmarkMethodForTest(t *testing.T, methodString string, p transport.Proto
...
-		method         string
-		wantErr        string
+		hostPort string
+		method   string
+		wantErr  string
 	}{
 		{
 			hostPort: s.hostPort(),
 		},
-		{
```
Comment (prashantv) [inline bench_method_test.go]: this looks unrelated to headers -- should this have been handled in a previous diff?

File: server_util_test.go [hunk 16]
```diff
@@ -37,8 +39,10 @@ type server struct {
 
 type handler func(ctx context.Context, args *raw.Args) (*raw.Res, error)
 
-func newServer(t *testing.T) *server {
-	ch := testutils.NewServer(t, testutils.NewOpts().SetServiceName("foo").DisableLogVerification())
+func newServer(t *testing.T, tracer opentracing.Tracer) *server {
```
Comment (prashantv) [inline server_util_test.go]: Rather than take a tracer (and get into the pattern of everyone passing `nil` for default options), we should probably just implement a lightweight options pattern here,


```go
func withTracer(opentracing.Tracer) func(*testutils.ChannelOpts) {
  return func(opts *testutils.ChannelOpts) {
    opts.Tracer = tracer
  }
}

func newServer(t *testing.T, opts... func(*testutils.ChannelOpts)) {
  chOpts := testutils.NewOpts().[...]
  for _, opt := range opts {
    opt(chOpts)
  }
  [...]
}
```

File: server_util_test.go [hunk 17]
```diff
@@ -78,11 +83,21 @@ func (methodsT) echo() handler {
...
+		if span != nil {
+			if spanCtx, ok := span.Context().(jaeger.SpanContext); ok {
+				if spanCtx.IsDebug() {
+					ret = 0x01
+				}
+			} else {
+				ret = 0xfe
+			}
+		} else {
+			ret = 0xff
```
Comment (prashantv) [inline server_util_test.go]: no span in context case returning 0 seems like a good default

File: main.go (hunk 9 above)
Comment (prashantv) [inline main.go]: yeah looks like the diff needs to be rebased off the latest `dev` branch

File: request_test.go [hunk 18]
```diff
@@ -141,7 +141,7 @@ func TestGetHeaders(t *testing.T) {
 	}
 
 	for _, tt := range tests {
-		got, err := getHeaders(tt.inline, tt.file)
+		got, err := getHeaders(tt.inline, tt.file, nil)
```
Comment (prashantv) [inline request_test.go]: can we add unit tests for the `override` here

File: transport/tchannel_test.go [hunk 19]
```diff
@@ -174,7 +174,6 @@ func TestTChannelCallSuccessRaw(t *testing.T) {
 	for _, tt := range tests {
 		var lastSpan uint64
 		testutils.RegisterFunc(svr, "echo", func(ctx context.Context, args *raw.Args) (*raw.Res, error) {
-			// assert.False(t, tchannel.CurrentSpan(ctx).TracingEnabled(), "Tracing should be disabled")
```
Comment (prashantv) [inline transport/tchannel_test.go]: intentional change? shouldn't this just be uncommenting?

File: transport/tchannel.go [hunk 20]
```diff
@@ -105,20 +112,30 @@ func NewTChannel(opts TChannelOptions) (Transport, error) {
...
+	return t.tracer
+}
+
 func (t *tchan) Protocol() Protocol {
 	return TChannel
 }
 
 func (t *tchan) Call(ctx context.Context, r *Request) (*Response, error) {
-	call, err := t.sc.BeginCall(ctx, r.Method, t.callOptions)
+	req := *r
```
Comment (prashantv) [inline transport/tchannel.go]: can you add a comment for why we do this

File: main.go [hunk 21]
```diff
@@ -309,6 +306,23 @@ type noEnveloper interface {
 	WithoutEnvelopes() encoding.Serializer
 }
 
+func getTracer(opts Options, out output) (opentracing.Tracer, io.Closer) {
+	var tracer opentracing.Tracer
+	var closer io.Closer
+	tracer = opentracing.NoopTracer{}
```
Comment (abhinav) [inline main.go]: nit: use var group and initialize tracer at declaration:

```
var (
  tracer opentracing.Tracer = opentracing.NoopTracer{}
  closer io.Closer
)
```

File: server_util_test.go (hunk 17 above)
Comment (kriskowal) [inline server_util_test.go]: these cases were helpful in isolating test errors during development. I’ll scale it back down.

File: bench_method_test.go (hunk 15 above)
Comment (kriskowal) [inline bench_method_test.go]: This is a far flung effect of refactoring tracing. Tracing is no longer a transport level concern, since we no longer use trace sampling on the channel level. Consequently it becomes a request scoped concern and the logic to detect whether the caller was overridden had to be factored upward so I could validate benchmark options.

File: transport/interface.go (hunk 11 above)
Comment (kriskowal) [inline transport/interface.go]: ✔

File: bench_method_test.go (hunk 15 above)
Comment (prashantv) [inline bench_method_test.go]: Ahh I see

File: integration_test.go [hunk 22]
```diff
@@ -80,9 +82,64 @@ func argHandler(arg int32) (int32, error) {
...
+		return errors.New("trace context is not from jaeger")
+	}
+	if !spanCtx.IsDebug() {
+		return errors.New("span context is not configured to submit traces")
+	}
+
+	val := span.BaggageItem("baggagekey")
+	if val == "" {
+		return errors.New("missing baggage")
+	} else if val != "baggagevalue" {
```
Comment (prashantv) [inline integration_test.go]: nit: no need for `else` since the above case returns,
```go
val := span.BaggageItem(..)
if val == "" {
  return errors.New("missing baggage")
}
if val != "baggagevalue" {
  return errors.New("unexpected baggage")
}
return nil
```

File: integration_test.go [hunk 23]
```diff
@@ -80,9 +82,64 @@ func argHandler(arg int32) (int32, error) {
...
+		return errors.New("unexpected baggage")
+	}
+	return nil
+}
+
+func verifyThriftHeaders(ctx thrift.Context) error {
+	headers := ctx.Headers()
+	if val, ok := headers["headerkey"]; ok {
```
Comment (prashantv) [inline integration_test.go]: nit: i'd write this the same as the `verifyBaggage` to reduce nesting
```go
val, ok := ctx.Headers()["headerkey"]
if !ok {
  return errors.New("missing header")
}
if val != "headervalue" {
  return errors.New("unexpected header")
}
return nil
```

File: integration_test.go [hunk 24]
```diff
@@ -80,9 +82,64 @@ func argHandler(arg int32) (int32, error) {
...
+		}
+	} else {
+		return errors.New("missing header")
+	}
+
+	return nil
+}
+
+func verifyYARPCHeaders(reqMeta yarpc.ReqMeta) error {
+	headers := reqMeta.Headers()
```
Comment (prashantv) [inline integration_test.go]: nit: can we instead have a single `veriyHeaders(headers map[string]string) error` method, and pass in the `ctx.Headers()` or `reqMeta.Headers()`? That way we can avoid the duplication here

File: transport/tchannel.go [hunk 25]
```diff
@@ -75,7 +81,7 @@ func NewTChannel(opts TChannelOptions) (Transport, error) {
 	}
 
 	callerName := opts.SourceService
-	if cn, ok := opts.TransportOpts["cn"]; ok {
+	if cn, ok := opts.TransportOpts["cn"]; ok && cn != "" {
```
Comment (prashantv) [inline transport/tchannel.go]: nit: should we keep this as is -- it might be we want to test what happens if the `cn` header is empty in a request.

File: integration_test.go (hunk 24 above)
Comment (kriskowal) [inline integration_test.go]: YARPC does not expose Headers().Items().

Comment (prashantv) [inline integration_test.go]: Ahh, I forgot that yarpc headers are not an actual map, this is fine then.
